### PR TITLE
Fix get_ip failed when doesn't install ifconfig

### LINKED ISSUE #12: util.sh中的get_ip函数存在bug
你好，在使用hugegraph-tools进行hugegraph server和studio的deploy过程中，我发现，有时候`rest-server.yml`和`hugegraph-studio.properties`文件能够被正确修改主机ip字段，有时候则为空。

```
function get_ip() {
    local os=`uname`
    local loopback="127.0.0.1"
    local ip=""
    case $os in
        Linux) ip=`ifconfig | grep 'inet addr:'| grep -v "$loopback" | cut -d: -f2 | awk '{ print $1}'`;;
        FreeBSD|OpenBSD|Darwin) ip=`ifconfig  | grep -E 'inet.[0-9]' | grep -v "$loopback" | awk '{ print $2}'`;;
        SunOS) ip=`ifconfig -a | grep inet | grep -v "$loopback" | awk '{ print $2} '`;;
        *) ip=$loopback;;
    esac
    echo $ip
}
```

### PR DESCRIPTION
**Pull request by Linary**
Fix #12

Change-Id: I2ae4e4e6e3d1749981155e477887a8a4269a26ba

### CODE REVIEW

File: assembly/bin/util.sh [hunk 1]
```diff
@@ -195,14 +195,48 @@ function ensure_path_writable() {
     fi
 }
 
+function command_available() {
+    local cmd=$1
+    if [ `command -v $cmd >/dev/null 2>&1` ]; then
+        echo "command $cmd is available"
```
Comment (zhoney) [inline assembly/bin/util.sh]: if exist, no need to echo info

File: assembly/bin/util.sh [hunk 2]
```diff
@@ -195,14 +195,48 @@ function ensure_path_writable() {
...
+        FreeBSD|OpenBSD|Darwin)
+            if [[ `command_available "ifconfig"` ]]; then
+                ip=`ifconfig  | grep -E 'inet.[0-9]' | grep -v "$loopback" | awk '{ print $2}'`
+            else
+                echo "Required ifconfig but it is not installed"
+                exit 1
+            fi
+            ;;
+        SunOS)
+            if [[ `command_available ifconfig` ]]; then
```
Comment (zhoney) [inline assembly/bin/util.sh]: "ifconfig"

File: assembly/bin/util.sh [hunk 3]
```diff
@@ -195,14 +195,48 @@ function ensure_path_writable() {
...
+            elif [[ `command_available "ip"` ]]; then
+                ip=`ip addr | grep 'state UP' -A2 | tail -n1 | awk '{print $2}' | awk -F"/" '{print $1}'`
+            else
+                echo "Required ifconfig or ip but they are not installed"
+                exit 1
+            fi
+            ;;
+        FreeBSD|OpenBSD|Darwin)
+            if [[ `command_available "ifconfig"` ]]; then
+                ip=`ifconfig  | grep -E 'inet.[0-9]' | grep -v "$loopback" | awk '{ print $2}'`
```
Comment (zhoney) [inline assembly/bin/util.sh]: two space after "ifconfig"

File: assembly/bin/util.sh [hunk 4]
```diff
@@ -195,14 +195,48 @@ function ensure_path_writable() {
...
-        Linux) ip=`ifconfig | grep 'inet addr:'| grep -v "$loopback" | cut -d: -f2 | awk '{ print $1}'`;;
-        FreeBSD|OpenBSD|Darwin) ip=`ifconfig  | grep -E 'inet.[0-9]' | grep -v "$loopback" | awk '{ print $2}'`;;
-        SunOS) ip=`ifconfig -a | grep inet | grep -v "$loopback" | awk '{ print $2} '`;;
+        Linux)
+            if [[ `command_available "ifconfig"` ]]; then
+                ip=`ifconfig | grep 'inet addr:'| grep -v "$loopback" | cut -d: -f2 | awk '{ print $1}'`
+            elif [[ `command_available "ip"` ]]; then
+                ip=`ip addr | grep 'state UP' -A2 | tail -n1 | awk '{print $2}' | awk -F"/" '{print $1}'`
+            else
+                echo "Required ifconfig or ip but they are not installed"
```
Comment (zhoney) [inline assembly/bin/util.sh]: use "unavailable" instead of "not installed"

**Review by zhoney — CHANGES_REQUESTED** [review]

File: assembly/bin/util.sh [hunk 5]
```diff
@@ -195,14 +195,48 @@ function ensure_path_writable() {
...
-        Linux) ip=`ifconfig | grep 'inet addr:'| grep -v "$loopback" | cut -d: -f2 | awk '{ print $1}'`;;
-        FreeBSD|OpenBSD|Darwin) ip=`ifconfig  | grep -E 'inet.[0-9]' | grep -v "$loopback" | awk '{ print $2}'`;;
-        SunOS) ip=`ifconfig -a | grep inet | grep -v "$loopback" | awk '{ print $2} '`;;
+        Linux)
+            if command_available "ifconfig"; then
+                ip=`ifconfig | grep 'inet addr:'| grep -v "$loopback" | cut -d: -f2 | awk '{ print $1}'`
+            elif command_available "ip"; then
+                ip=`ip addr | grep 'state UP' -A2 | tail -n1 | awk '{print $2}' | awk -F"/" '{print $1}'`
+            else
+                echo "Required ifconfig or ip but they are available"
```
Comment (zhoney) [inline assembly/bin/util.sh]: unavailable

File: assembly/bin/util.sh [hunk 6]
```diff
@@ -195,14 +195,48 @@ function ensure_path_writable() {
...
+                exit 1
+            fi
+            ;;
+        FreeBSD|OpenBSD|Darwin)
+            if command_available "ifconfig"; then
+                ip=`ifconfig | grep -E 'inet.[0-9]' | grep -v "$loopback" | awk '{ print $2}'`
+            else
+                echo "Required ifconfig but it is available"
```
Comment (zhoney) [inline assembly/bin/util.sh]: ditto

File: assembly/bin/util.sh [hunk 7]
```diff
@@ -195,14 +195,48 @@ function ensure_path_writable() {
...
+                exit 1
+            fi
+            ;;
+        SunOS)
+            if command_available "ifconfig"; then
+                ip=`ifconfig -a | grep inet | grep -v "$loopback" | awk '{ print $2} '`
+            else
+                echo "Required ifconfig but it is available"
```
Comment (zhoney) [inline assembly/bin/util.sh]: ditto

**Review by zhoney — CHANGES_REQUESTED** [review]

**Review by zhoney — DISMISSED** [review]

File: assembly/bin/util.sh [hunk 8]
```diff
@@ -195,14 +195,46 @@ function ensure_path_writable() {
...
-        FreeBSD|OpenBSD|Darwin) ip=`ifconfig  | grep -E 'inet.[0-9]' | grep -v "$loopback" | awk '{ print $2}'`;;
-        SunOS) ip=`ifconfig -a | grep inet | grep -v "$loopback" | awk '{ print $2} '`;;
+        Linux)
+            if command_available "ifconfig"; then
+                ip=`ifconfig | grep 'inet addr:' | grep -v "$loopback" | cut -d: -f2 | awk '{ print $1}'`
+            elif command_available "ip"; then
+                ip=`ip addr | grep 'state UP' -A2 | tail -n1 | awk '{print $2}' | awk -F"/" '{print $1}'`
+            else
+                echo "Required ifconfig or ip but they are unavailable"
+                exit 1
```
Comment (javeme) [inline assembly/bin/util.sh]: prefer to set: ip=$loopback;;

**Review by javeme — CHANGES_REQUESTED** [review]

**Review by javeme — APPROVED** [review]

**Review by zhoney — APPROVED** [review]

**Pull request merged**